### ISSUE TITLE
upload asset to github release?

### ISSUE THREAD

**Issue by topic2k** [mozilla/agithub#20]
How do i upload an asset to a release?
With the following code i get `socket.error: [Errno 10054]` (connection closed by remote).

```
import base64
from agithub import GitHub

with open('MyApp.exe', 'rb') as f:
    body = f.read()

gh = Github(token=token)
rc, data = gh.repos[user][repo].releases[releaseId].assets.post(
    url= uploadUrl,
    name='MyApp.exe',
    body=base64.encodestring(body),
    headers={'Content-Type': 'application/octet-stream'}
    )
```

If i don't encode to base64 i get an UnicodeDecodeError.

What is the correct way to do it?
I'm using `Python 2.7.9 Stackless 3.1b3 060516 (default, Feb 21 2015, 11:54:09) [MSC v.1500 32 bit (Intel)] on win32` and release 2.0 of agithub.

**Comment by jpaugh** [comment]
The uploading code is new and mostly untested. It's probably broken. I'll look into it.

**Comment by topic2k** [comment]
I took some time and found a solution:
The GitHub Client is created with `api_url = 'api.github.com'`, but for asset uploads it must be `api_url = 'uploads.github.com'`. So i extended/changed my above code to:

```
from agithub.GitHub import GitHub
from agithub.base import ConnectionProperties

with open('MyApp.exe', 'rb') as f:
    file_content = f.read()

conn_props = ConnectionProperties(
    api_url = "uploads.github.com",
    secure_http = True,
    extra_headers = {
        'authorization' : 'Token 8ed05...',
        'accept': 'application/vnd.github.v3+json'
    }
)
gh = GitHub(token=token, connection_properties=conn_props)

rc, data = gh.repos[user][repo].releases[releaseId].assets.post(
    url = uploadUrl,
    name = 'MyApp.exe',
    body = file_content,
    headers = {'content-type': 'application/octet-stream'}
    )
```

And in `base.py` `Client` i changed `def post ...` to:

```
def post(self, url, body=None, headers={}, **params):
    url += self.urlencode(params)
    if not 'content-type' in headers or headers.get('content-type') == 'application/json':
        # We're doing a json.dumps of body, so let's set the content-type to json
        headers['content-type'] = 'application/json'
        return self.request('POST', url, json.dumps(body), headers)
    else:
        return self.request('POST', url, body, headers)
```

Now the upload of the asset works, but i'm not sure if the change of `post` is the right way or how to handle it in another way.

**Comment by jpaugh** [comment]
Thanks for looking into this! I'm glad to see `agithub` is getting some use. It looks like you're using an older version of the library. Try the latest and see if that helps.

In particular, the GitHub client now allows you to pass in an arbitrary `api_url`. Also, `post`, etc. rely on `RequestBody` to do any meda-type specific handling. Take a look at [`RequestBody.process`](https://github.com/jpaugh/agithub/blob/master/agithub/base.py#L353), and where it is called in [`Client.request`](https://github.com/jpaugh/agithub/blob/master/agithub/base.py#L194).

**Comment by topic2k** [comment]
Well, i used the latest [release ](https://github.com/jpaugh/agithub/releases/tag/v2.0). Now i downloaded the master branch and tested again and it's working (with a small fix #23). My code can now be shortend to:

```
from agithub.GitHub import GitHub

with open('MyApp.exe', 'rb') as f:
    file_content = f.read()

gh = GitHub(token=token, api_url = "uploads.github.com")

rc, data = gh.repos[user][repo].releases[releaseId].assets.post(
    url = uploadUrl,
    name = 'MyApp.exe',
    body = file_content,
    headers = {'content-type': 'application/octet-stream'}
    )
```

Maybe it's time for a new release?

**Comment by jpaugh** [comment]
You're absolutely right! I'll close up the 2.1 development branch.
